### PR TITLE
feat(server): AWS Step Functions MCP Server (#298)

### PR DESCRIPTION
**Pull request by mmouniro**
Fixes

## Summary

This change implements a new MCP server that acts as a bridge between MCP clients and AWS Step Functions, enabling foundation models to execute and interact with Step Function workflows as tools. This builds upon the pattern established by the Lambda MCP server, extending similar capabilities to Step Functions.

### Changes

> Please provide a summary of what's being changed

### User experience

> Please share what the user experience looks like before and after this change

See https://github.com/awslabs/mcp/issues/298

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

Is this a breaking change? (Y/N)

**RFC issue number**: #298 

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

### CODE REVIEW

**Review by awslabs-mcp — COMMENTED** [review]
## General Assessment

I've reviewed the PR #377 implementing the new AWS Step Functions MCP Server. This server follows the pattern established by the Lambda MCP server, extending similar capabilities to Step Functions. 

Key observations:
- The implementation is well-structured and comprehensive, following the MCP server design guidelines
- Documentation is thorough and clearly explains usage patterns
- Test coverage is quite good at 97.31%, with only 5 partial lines lacking coverage
- The server properly handles both Standard and Express workflow types
- Integration with EventBridge Schema Registry for input validation is a nice feature

The implementation uses proper AWS authentication patterns, error handling, and follows the project's coding standards. The examples provided are clear and illustrative of the server's capabilities.

Workflow Run ID: 15149872188

**Review by awslabs-mcp — COMMENTED** [review]
## File-Specific Comments

### 

1. **State Machine Enumeration**: The implementation properly filters state machines based on name prefix, explicit list, and tags, providing good security boundaries.
  
2. **Schema Integration**: The integration with EventBridge Schema Registry is well-implemented, allowing for formal validation of inputs.
  
3. **Polling Implementation**: For Standard workflows, the polling implementation is clean, though it might benefit from a configurable timeout to prevent indefinite polling.

4. **Documentation Generation**: The approach of combining state machine description, workflow comment, and input schema into comprehensive tool documentation is excellent.

### 

Very well documented with clear instructions on setup, configuration, and security considerations. The mermaid diagram provides a helpful visual overview of the architecture.

### Tests

Test coverage is comprehensive, covering the main functionality paths and edge cases. The tests for both Standard and Express workflows are thorough.

### Example State Machines

The example state machines demonstrate the server's capabilities well, including schema validation. The template.yml is particularly helpful in showing how to configure state machines for use with this MCP server.

Workflow Run ID: 15149872188

**Review by awslabs-mcp — COMMENTED** [review]
## File-Specific Comments

### server.py

1. **State Machine Enumeration**: The implementation properly filters state machines based on name prefix, explicit list, and tags, providing good security boundaries.
  
2. **Schema Integration**: The integration with EventBridge Schema Registry is well-implemented, allowing for formal validation of inputs.
  
3. **Polling Implementation**: For Standard workflows, the polling implementation is clean, though it might benefit from a configurable timeout to prevent indefinite polling.

4. **Documentation Generation**: The approach of combining state machine description, workflow comment, and input schema into comprehensive tool documentation is excellent.

### README.md

Very well documented with clear instructions on setup, configuration, and security considerations. The mermaid diagram provides a helpful visual overview of the architecture.

### Tests

Test coverage is comprehensive, covering the main functionality paths and edge cases. The tests for both Standard and Express workflows are thorough.

### Example State Machines

The example state machines demonstrate the server's capabilities well, including schema validation. The template.yml is particularly helpful in showing how to configure state machines for use with this MCP server.

Workflow Run ID: 15149872188

**Review by awslabs-mcp — APPROVED** [review]
## Final Assessment

After a thorough review of PR #377 which implements the AWS Step Functions MCP Server, I'm approving this pull request. The implementation follows the established patterns for MCP servers in this repository and properly extends the capabilities to AWS Step Functions.

### Key Strengths:
- Well-structured code following repository patterns and best practices
- Excellent documentation with clear usage examples
- Strong test coverage (97.31%)
- Proper security boundaries with state machine filtering
- Integration with EventBridge Schema Registry for input validation
- Support for both Standard and Express workflows
- Good error handling and logging

### Minor Improvement Suggestions (for future enhancements):
1. Consider adding a configurable timeout for Standard workflow polling to prevent indefinite polling
2. Consider adding more examples covering error cases and complex workflow patterns

These suggestions are minor and can be addressed in future PRs. The current implementation is ready for merging.

Workflow Run ID: 15149872188

**Review by awslabs-mcp — APPROVED** [review]
# AWS Step Functions MCP Server Review\n\nThank you for implementing this MCP server for AWS Step Functions. I've reviewed the code in detail and have the following feedback.\n\n## Overall Assessment\n\nThe implementation follows best practices from the existing Lambda MCP server pattern while adapting it appropriately for Step Functions workflows. The code is well-structured, includes comprehensive documentation, and has excellent test coverage.\n\n### Strengths\n\n- Clear, comprehensive README with detailed usage instructions\n- Support for both Standard and Express state machine types with appropriate execution patterns\n- Integration with EventBridge Schema Registry for input validation\n- Good error handling and logging practices\n- Thorough test coverage (97.31% according to Codecov report)\n- Well-documented example state machines\n- Strong security considerations described in the README\n\n### Minor Improvement Suggestions\n\n1. Error messages in the  and  functions could be more descriptive. Consider adding more context about the error cause.\n\n2. There's a TODO or improvement opportunity in relation to the Codecov report showing 5 lines that need coverage. These appear to be partial branches in the error handling code.\n\n3. Consider adding timeout parameters to both the standard and express state machine execution functions to prevent indefinite waiting, especially for the polling loop in the standard execution implementation.\n\n## Recommendation\n\nI recommend **APPROVING** this pull request with minor suggestions for future improvements. The implementation is solid, follows established patterns, and will be a valuable addition to the MCP ecosystem.\n\nReview by Claude Code with run ID: 15167624114

**Review by awslabs-mcp — APPROVED** [review]
# AWS Step Functions MCP Server Review

Thank you for implementing this MCP server for AWS Step Functions. I've reviewed the code in detail and have the following feedback.

## Overall Assessment

The implementation follows best practices from the existing Lambda MCP server pattern while adapting it appropriately for Step Functions workflows. The code is well-structured, includes comprehensive documentation, and has excellent test coverage.

### Strengths

- Clear, comprehensive README with detailed usage instructions
- Support for both Standard and Express state machine types with appropriate execution patterns
- Integration with EventBridge Schema Registry for input validation
- Good error handling and logging practices
- Thorough test coverage (97.31% according to Codecov report)
- Well-documented example state machines
- Strong security considerations described in the README

### Minor Improvement Suggestions

1. Error messages in the `invoke_standard_state_machine_impl` and `invoke_express_state_machine_impl` functions could be more descriptive. Consider adding more context about the error cause.

2. There's a TODO or improvement opportunity in relation to the Codecov report showing 5 lines that need coverage. These appear to be partial branches in the error handling code.

3. Consider adding timeout parameters to both the standard and express state machine execution functions to prevent indefinite waiting, especially for the polling loop in the standard execution implementation.

## Recommendation

I recommend **APPROVING** this pull request with minor suggestions for future improvements. The implementation is solid, follows established patterns, and will be a valuable addition to the MCP ecosystem.

Review by Claude Code with run ID: 15167624114

**Review by scottschreckengaust — CHANGES_REQUESTED** [review]
Thank you for the contribution. Please move the sample into the `/samples`

**Comment by awslabs-mcp** [comment]
## Step Functions MCP Server Implementation Review

Overall, this PR implements a new MCP server that acts as a bridge between MCP clients and AWS Step Functions. The implementation follows established patterns from the Lambda MCP server, which is a good approach for consistency.

### Strong Points
1. The server properly supports both Standard and Express workflows with appropriate implementation differences.
2. The EventBridge Schema Registry integration provides valuable input validation.
3. Documentation is thorough with clear examples and security considerations.
4. Tests are comprehensive, covering both success and error cases.
5. The server handles various filtering options for state machines (prefix, list, tags).

### Items Noted
1. The fail-by-label check is failing, but this appears to be unrelated to the code quality.
2. The server properly implements the design guidelines from DESIGN_GUIDELINES.md.
3. Authentication handling follows best practices from other MCP servers.

The Pull Request workflow check (Claude) is still pending, but this shouldn't block the review as instructed.

**Review by awslabs-mcp — APPROVED** [review]
I've reviewed the AWS Step Functions MCP Server implementation. The code is well-structured, follows the established MCP server patterns, and includes proper documentation and tests. The server correctly implements the functionality to register Step Functions state machines as MCP tools with support for both Standard and Express workflows, EventBridge Schema Registry integration, and various filtering options.

I believe this implementation is ready to be merged as it meets the project's quality standards and provides a valuable new capability to the MCP ecosystem. Run ID: 15168613118

**Comment by mmouniro** [comment]
> Thank you for the contribution. Please move the sample into the `/samples`

done

File: src/stepfunctions-tool-mcp-server/awslabs/stepfunctions_tool_mcp_server/server.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    args = parser.parse_args()
+
+    register_state_machines()
+
+    # Run server with appropriate transport
+    if args.sse:
+        mcp.settings.port = args.port
+        mcp.run(transport='sse')
+    else:
+        mcp.run()
```
Comment (alexa-perlov) [inline src/stepfunctions-tool-mcp-server/awslabs/stepfunctions_tool_mcp_server/server.py]: We're no longer supporting SSE in this repo 

```suggestion
def main():
    """Run the MCP server."""
    register_state_machines()
    mcp.run()
```

**Review by awslabs-mcp — APPROVED** [review]
# AWS Step Functions MCP Server Review

## Initial Assessment

I have thoroughly reviewed PR #377 which implements a new AWS Step Functions MCP Server. This server acts as a bridge between MCP clients and AWS Step Functions, enabling foundation models to execute state machines as tools.

## Key Observations

### Strengths

1. **Design Consistency**: The implementation follows the established pattern from the Lambda MCP server and aligns with the repository's design guidelines from DESIGN_GUIDELINES.md.

2. **Comprehensive Implementation**: 
   - Supports both Standard and Express workflows with appropriate execution patterns
   - Integration with EventBridge Schema Registry for input validation
   - Robust filtering mechanisms (prefix, list, and tag-based)
   - Proper error handling and logging

3. **Documentation Quality**: 
   - Clear README with detailed setup instructions
   - Mermaid diagram explaining the architecture
   - Comprehensive examples in the samples directory
   - Good docstrings throughout the code

4. **Test Coverage**: Excellent test coverage at 97.31% with comprehensive test files covering all major functionality.

5. **Security Considerations**: Proper IAM-based filtering and validation mechanisms to ensure only authorized state machines can be executed.

### Technical Review

The server implementation in  is well-structured:
- Clean separation of concerns between Standard and Express execution patterns
- Proper async handling for state machine polling
- Good integration with AWS APIs (Step Functions and EventBridge Schema Registry)
- Appropriate use of environment variables for configuration

### Code Quality

- Follows repository coding standards (ruff, pyright)
- Consistent naming conventions
- Proper error handling with contextual error messages
- Good logging practices

## Status Checks

All required checks are passing except for the  check (which appears unrelated to code quality) and the pending  workflow (which we're instructed not to wait for per guidelines).

## Issue Resolution

I note that scottschreckengaust requested moving the sample to , which has been completed as confirmed by the author's response.

## Final Assessment

This is a high-quality implementation that adds valuable functionality to the MCP ecosystem. The code follows established patterns, includes comprehensive testing, and provides clear documentation. The implementation is ready for production use.

**RECOMMENDATION: APPROVE**

The implementation meets all project quality standards and will be a valuable addition to the AWS MCP server collection.

Review by Claude Code with run ID: 15308429589

**Review by awslabs-mcp — APPROVED** [review]
# AWS Step Functions MCP Server Review

## Initial Assessment

I have thoroughly reviewed PR #377 which implements a new AWS Step Functions MCP Server. This server acts as a bridge between MCP clients and AWS Step Functions, enabling foundation models to execute state machines as tools.

## Key Observations

### Strengths

1. **Design Consistency**: The implementation follows the established pattern from the Lambda MCP server and aligns with the repository design guidelines.

2. **Comprehensive Implementation**: 
   - Supports both Standard and Express workflows with appropriate execution patterns
   - Integration with EventBridge Schema Registry for input validation
   - Robust filtering mechanisms (prefix, list, and tag-based)
   - Proper error handling and logging

3. **Documentation Quality**: 
   - Clear README with detailed setup instructions
   - Mermaid diagram explaining the architecture
   - Comprehensive examples in the samples directory
   - Good docstrings throughout the code

4. **Test Coverage**: Excellent test coverage at 97.31% with comprehensive test files covering all major functionality.

5. **Security Considerations**: Proper IAM-based filtering and validation mechanisms to ensure only authorized state machines can be executed.

### Code Quality

- Follows repository coding standards
- Consistent naming conventions
- Proper error handling with contextual error messages
- Good logging practices

## Final Assessment

This is a high-quality implementation that adds valuable functionality to the MCP ecosystem. The code follows established patterns, includes comprehensive testing, and provides clear documentation.

**RECOMMENDATION: APPROVE**

Review by Claude Code with run ID: 15308429589

File: src/stepfunctions-tool-mcp-server/tests/test_main.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        # Set up test data
+        mock_parse_args.return_value = MagicMock(sse=True, port=8888)
+
+        # Call the function
+        main()
+
+        # Verify results
+        mock_register.assert_called_once()
+        mock_mcp.run.assert_called_once_with(transport='sse')
+        assert mock_mcp.settings.port == 8888
```
Comment (alexa-perlov) [inline src/stepfunctions-tool-mcp-server/tests/test_main.py]: Remove because no longer supporting SSE

File: src/stepfunctions-tool-mcp-server/tests/test_main.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        custom_port = 9999
+        mock_parse_args.return_value = MagicMock(sse=True, port=custom_port)
+
+        # Call the function
+        main()
+
+        # Verify results
+        mock_register.assert_called_once()
+        mock_mcp.run.assert_called_once_with(transport='sse')
+        assert mock_mcp.settings.port == custom_port
```
Comment (alexa-perlov) [inline src/stepfunctions-tool-mcp-server/tests/test_main.py]: Remove because no longer supporting SSE

File: src/stepfunctions-tool-mcp-server/tests/test_main.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        # Set up test data
+        mock_parse_args.return_value = MagicMock(sse=True, port=8888)  # Default port
+
+        # Call the function
+        main()
+
+        # Verify results
+        mock_register.assert_called_once()
+        mock_mcp.run.assert_called_once_with(transport='sse')
+        assert mock_mcp.settings.port == 8888
```
Comment (alexa-perlov) [inline src/stepfunctions-tool-mcp-server/tests/test_main.py]: Remove because no longer supporting SSE

File: src/stepfunctions-tool-mcp-server/tests/test_main.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            mock_register.assert_called()
+            if args['sse']:
+                mock_mcp.run.assert_called_with(transport='sse')
+                assert mock_mcp.settings.port == args['port']
+            else:
+                mock_mcp.run.assert_called_with()
+
+            # Reset mocks for next iteration
+            mock_register.reset_mock()
+            mock_mcp.reset_mock()
```
Comment (alexa-perlov) [inline src/stepfunctions-tool-mcp-server/tests/test_main.py]: remove because no longer supporting SSE

File: src/stepfunctions-tool-mcp-server/tests/test_main.py (hunk 3 above)
Comment (mmouniro) [reply src/stepfunctions-tool-mcp-server/tests/test_main.py]: fixed

File: src/stepfunctions-tool-mcp-server/tests/test_main.py (hunk 4 above)
Comment (mmouniro) [reply src/stepfunctions-tool-mcp-server/tests/test_main.py]: fixed

**Review by scottschreckengaust — APPROVED** [review]
LGTM
